Relationship mapping

snusnu opened the pull request:
This is not yet complete obviously, but this PR will serve as a place for more discussion while fleshing out the details.

snusnu committed:
Initial code for relationship mapper generation

snusnu committed:
Support setting Mapper.model more than once

snusnu committed:
Initial code for supporting autogenerated 1:1 mappers

snusnu committed:
Define extra mapping for 1:1 integration spec

snusnu committed:
Revert "Remove Inflector"

This reverts commit ea4e35095332c1ae6d9282ae736e7fa4f6e51ae8.

snusnu committed:
DRY up by introducing Relationship::Options

This commit also uses the previously re-added
inflector code to provide the default :target_key
option for 1:1, 1:N and N:1 relationships.

snusnu committed:
Fix typo in one_to_many integration spec

snusnu committed:
Cleanup whitespace

solnic commented:
Looks like it fails only on ruby 1.8.7

snusnu committed:
Initial support for autogenerated 1:1 mappers

snusnu committed:
Support changing Mapper.relation_name

snusnu committed:
Support changing Mapper.repository

snusnu committed:
Fix a couple of spec failures on 1.8.7

snusnu committed:
Workaround a problem with Mapper.inherited on 1.8.7

snusnu committed:
Remove previously introduced debug code

snusnu commented:
I fixed the spec failures on 1.8.7. It seems like 1.8.7 doesn't like my `Mapper.inherited`. Somehow, the `model` class isn't properly inherited or "resetted". See 7b37438

snusnu committed:
Initial support for autogenerated M:N mappers

snusnu committed:
Only rename existing attributes in M:N join relation

snusnu committed:
Add integration specs for all kinds of relationships

snusnu committed:
Introduce Relationship::ManyToMany#finalize_aliases

snusnu committed:
Update DataMapper::Mapper.unique_alias signature

snusnu committed:
Finalize aliases for all types of relationships

snusnu committed:
Make the receiver of the call explicit

snusnu committed:
Make a formatting change

snusnu committed:
Rename Relationship::Options#renamings to #aliases

snusnu committed:
Temporarily remove jruby-head from travis setup

jruby-head on travis is currently missing bundler

dkubb commented on lib/data_mapper/mapper/relationship/many_to_many.rb:
```diff
@@ -8,13 +8,37 @@ class Relationship
+
+          @join_aliases = Hash[via_key.zip(target_key)]
+          target_key.each do |attribute_name|
+            if (@join_relation.header[attribute_name] rescue false)
```
Should I provide a way in `Veritas::Relation::Header` to test if an attribute exists?

dkubb commented on lib/data_mapper/mapper/relationship/options.rb:
```diff
@@ -0,0 +1,68 @@
+        def initialize(name, source_model, target_model, options)
+          @name         = name.to_sym
+          @options      = options
+          @source       = options[:source]
```
It's a small point, but I wonder if you want to use `Hash#fetch` anytime you're accessing something in a Hash that you expect to exist. These days I only ever use `Hash#[]` when it's ok for the value to be nil.

dkubb commented on lib/data_mapper/mapper/relationship/options.rb:
```diff
@@ -0,0 +1,68 @@
+            @target_model = @mapper_class.attributes[@name].type
+          else
+            @source_model = source_model
+            @target_model = target_model || options.delete(:model)
```
I would be careful using `options.delete(:model)` here. This code modifies the arguments to the method, and if you're using the object elsewhere, it's now been modified. In general I think it's better to assume every argument to a method is immutable; unless the method's purpose is to mutate the argument (which is really, really rare in good OO code).

Another thing I would suggest is looking at the immutable stuff in Veritas. I've found some amazing benefits working with immutable objects, and being forced to inject everything into the constructor, and having to explicitly "mutate" state by returning a new object. There's an article about this that's great if you haven't read it yet: http://www.harukizaemon.com/2010/03/functional-programming-in-object-oriented-languages.html

dkubb commented on lib/data_mapper/mapper/veritas_mapper.rb:
```diff
@@ -22,8 +26,8 @@ def self.base_relation
       # @api public
-      def initialize(relation)
-        @relation      = relation
+      def initialize(relation = nil)
```
What about:

``` ruby
def initialize(relation = self.class.relation)
  @relation = relation
  # ...
```

dkubb commented on spec/integration/generated_mappers/relationship_many_to_one_spec.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+
+      class Mapper < DataMapper::Mapper::VeritasMapper
+        map :id,   :type => Integer, :key => true
```
I think I've mentioned this to @solnic before, but if we're always specifying the type I wonder if it should become a position argument like it is in other related systems.

dkubb commented:
@snusnu wow, this is pretty amazing. I looked over the diff and mentioned a few things, but there's nothing major.

I had some ideas about the API that I thought would be good to discuss with everyone. Some things we talked about on IRC, but I'll mention them here to spark discussion.

What do you think about an API that uses collections for 1:1, 1:m and m:m associations? Right now in DM1 we kind of fake it by making 1:1 behave kind of like m:1. I would like to make it so that those return collections.

I think m:1 should be defined closer to how an attribute is. The ideal case is that it's nearly indistinguishable, so that an attribute, embedded value, and m:1 association are the same at the object level. If we could move a "slot" (for lack of a better term) back and forth between these three types with minimal changes at the model level, and none in the object code that would be awesome. It would allow people to choose one approach and refactor the code later on and extract an embedded value into a separate relation without breaking existing application code.

I wonder if it would be possible during finalization to inspect all the relationships, and then perform a rename on the attributes so that the attributes that don't join to any other relations are unique, and any attributes that do join together are all given the same unique name so that the relations can be natural joined anytime. To illustrate what I mean consider:

``` ruby
class Customer
  attr_reader :id, :name, :email

  def initialize(attributes)
    @id, @name, @email = attributes.values_at(:id, :name, :email)
  end

  class Mapper < DataMapper::Mapper::VeritasMapper
    model         Customer
    relation_name :customer
    repository    :postgres

    map :id,    :type => Integer, :key => true
    map :name,  :type => String
    map :email, :type => String

    has 1, :address, Address
    has n, :orders,  Order
  end
end
```

Now you can imagine the Address and Order objects both have customer_id columns. The system should be able to examine all the links into Customer and decide on a unique name for the fk that does not conflict with any other attribute in the system (perhaps prefixing it with the repository and relation name), and it would rename the "id" attribute underneath to use that attribute in the relation. It would make sure Address and Order use the same attribute name. From that point on you could perform the joins using `Customer::Mapper.relation.join(Address::Mapper.relation)` and it would just be a natural join. The mapper would map this unique attribute name back to the object address (in this case "id").

I often wonder if in a well designed database with good names, this kind of renaming wouldn't be necessary at all. It's too bad we latched onto a convention using "id" as the surrogate PK. If it was always "relation_name_id" or something, then we would make sure all FK's use the same name, and everything would just work.

dkubb commented on lib/data_mapper/mapper.rb:
```diff
@@ -30,21 +43,33 @@ def self.relation_registry
     # @api public
     def self.model(model=nil)
-      @model ||= model
+      if model
```
Usually what I would do in this case is have something like [this](https://github.com/dkubb/veritas/blob/master/lib/veritas.rb#L189). So this could become:

``` ruby
def self.model(model = Undefined)
  if model.equal?(Undefined)
    @model
  else
    @model = model
  end
end
```

dkubb commented:
This may be a bit "out there", but I've often wondered if what we call a relationship in DM1 could be modeled a bit differently.

If we had the concept of an edge (think graph theory) which specifies which nodes (relations) are combined into a binary op, which binary operation it is, and what attributes are being used. It doesn't specify anything about the direction, whether it's a 1:1 or 1:m on one side and a m:1 on the other, etc. There's no source or target, it's just two points that are used in the operation.

There would need to be a second object that is bound to a node that is the source. It would use the edge object to find the target. The relationship on the other side, would use the _same_ edge object, only it would be bound to the node on the other side. There should be a registry that only allows one distinct kind of edge between two points.

Well, the relationship wouldn't _have_ to be bidirectional, but if a bidirectional relationship did exist, it should share the same edge.

One advantage this gives is the ability to fairly easily identify the exact inverse of every relationship, a problem that was really difficult in DM1. I had cases where it was nearly impossible to figure out what the inverse was without a hint. We needed to know it to perform various kinds of queries. It also decomposes the problem of relationships into two parts: the edge or method of combining the relations, and the name/method of access on the objects.

This might end up being too weird, but if it helps model the system cleaner then I figured I should mention it.

snusnu committed:
Use parameter default value instead of a conditional

snusnu committed:
Tighten up the conditional guard in get/set methods

Makes it so that if a value that is not Undefined
gets passed in, it gets assigned to the ivar. If
no value (or Undefined) gets passed in, the ivar
will be returned.

Because of the use of DataMapper::Mapper.model in
DataMapper::Mapper.inherited, I had to move the
toplevel require statements below the toplevel
DataMapper module definition.

snusnu committed:
Call the attribute reader instead of using #[]

snusnu committed:
Use Hash#fetch instead of #[] if appropriate

snusnu committed:
Move type args extraction code into DataMapper::Utils

snusnu committed:
Make Mapper.map accept the type as the 2nd parameter

snusnu committed:
Set model/relation_name/repository before mappings

snusnu commented:
@dkubb I think I've picked all the low hanging fruits based on your comments. I will think some more about your suggestions concerning relationships and get back to you.

snusnu committed:
Update README to reflect the current mapper API

snusnu committed:
Rearrange method definitions in DataMapper::Mapper

snusnu committed:
Minor formatting changes

snusnu commented:
@dkubb re your question if `Veritas::Relation::Header` would benefit from a method that checks if a given attribute is present: I think it would. Initially I tried the following:

``` ruby
  if @join_relation.header.include?(attribute_name)
```

which lead to the following error:

```
.../veritas/lib/veritas/support/equalizer.rb:45:in `block (2 levels) in define_cmp_method': undefined method `size' for #<Veritas::Attribute::Object name=:id required?=true> (NoMethodError)
.../veritas/lib/veritas/support/equalizer.rb:45:in `each'
.../veritas/lib/veritas/support/equalizer.rb:45:in `all?'
.../veritas/lib/veritas/support/equalizer.rb:45:in `block in define_cmp_method'
.../veritas/lib/veritas/support/equalizer.rb:120:in `=='
.../veritas/lib/veritas/relation/header.rb:123:in `block in each'
.../veritas/lib/veritas/relation/header.rb:123:in `each'
.../veritas/lib/veritas/relation/header.rb:123:in `each'
.../dm-mapper/lib/data_mapper/mapper/relationship/many_to_many.rb:31:in `include?'
.../dm-mapper/lib/data_mapper/mapper/relationship/many_to_many.rb:31:in `block in finalize_aliases'
.../dm-mapper/lib/data_mapper/mapper/relationship/many_to_many.rb:29:in `each'
.../dm-mapper/lib/data_mapper/mapper/relationship/many_to_many.rb:29:in `finalize_aliases'
.../dm-mapper/lib/data_mapper/mapper/relationship.rb:34:in `finalize'
.../dm-mapper/lib/data_mapper/mapper/relationship/one_to_many.rb:48:in `finalize'
.../dm-mapper/lib/data_mapper/mapper/relationship_set.rb:24:in `block in finalize'
.../dm-mapper/lib/data_mapper/mapper/relationship_set.rb:18:in `block in each'
.../dm-mapper/lib/data_mapper/mapper/relationship_set.rb:18:in `each_value'
.../dm-mapper/lib/data_mapper/mapper/relationship_set.rb:18:in `each'
.../dm-mapper/lib/data_mapper/mapper/relationship_set.rb:24:in `finalize'
.../dm-mapper/lib/data_mapper/mapper.rb:108:in `finalize_relationships'
.../dm-mapper/lib/dm-mapper.rb:66:in `block in finalize'
.../dm-mapper/lib/dm-mapper.rb:66:in `each'
.../dm-mapper/lib/dm-mapper.rb:66:in `finalize'

```

snusnu committed:
Fix alignment and rearrange method calls in specs

snusnu committed:
Rearrange namespaces and thus directory layout

snusnu committed:
Distinguish between relation & base_relation mappers

snusnu committed:
Make generated mappers instances of Mapper::Relation 

Relying on the Mapper.inherited hook to generate
the new mapper would incorrectly make that mapper
a base relation mapper, at least if inherited from
a descendant of Mapper::Relation::Base.

snusnu committed:
Remove Mapper.inherited until a use-case shows up

snusnu committed:
Revert "Workaround a problem with Mapper.inherited on 1.8.7"

This reverts commit 7b374383aa448f6c217ac617417d8f36b679b07a.

The Mapper.inherited hook was previously removed

snusnu committed:
Depend on dkubb/descendants_tracker from github

Previously, descendants_tracker functionality was
provided from virtus, but dm-mapper itself should
not need to depend on virtus.

snusnu committed:
Remove Mapper.has_many in favor of Mapper.has n

snusnu committed:
Use DataMapper::Mapper::Relation::Base in the README

snusnu commented:
@dkubb I think I agree. I even wanted to just copy [extract_min_max](https://github.com/datamapper/dm-core/blob/master/lib/dm-core/model/relationship.rb#L236-244) from DM1. I skipped it for now tho, mainly because I have no idea yet about how to actually enforce those constraints. For the future, I'm all for it tho. Do you think we should start accepting ranges right now, and work on enforcing them later on?

dkubb commented:
@snusnu yeah, I would probably do that. We could also just coerce everything into a Range internally, and then at some point remove the coercion.  So for example:
- has n -> has 0..n
- has 1 -> has 1..1
- has 2 -> has 2..2

EDIT: I would never want to encourage people to actually use `1..1` as an argument, but we could coerce the value so it is represented that way internally.

snusnu committed:
Cardinality must either be a Range or an Integer

snusnu committed:
Add a link to the integration tests to the README

mbj commented on lib/data_mapper/relationship/one_to_many.rb:
```diff
@@ -0,0 +1,64 @@
+
+          parents = tuples.each_with_object({}) do |tuple, hash|
+            key = parent_key.map { |attribute| tuple[attribute.field] }
+            hash[key] = @attributes.primitives.each_with_object({}) { |attribute, parent|
```
As this is a multiline block I think the ``do ... end` form would be nicer here.

snusnu replied:
I _think_ the convention being used here is that if the block return value is being used, `{}` will be used as delimiters. I might be wrong tho.

solnic replied:
Yes that's the convention I'm following

snusnu committed:
Record the cardinality in Relationship::Options

snusnu committed:
The relationship source_model is always known

snusnu committed:
Rename Relationship#relationship_builder to #builder_class

snusnu committed:
Introduce Relationship::Options::Validator

The idea is to put option validation in those
dedicated objects. Every Relationship::Options
subclass will use its dedicated validator object
to perform validation.

snusnu committed:
Rename Relationship::Builder to Mapper::Builder::Relationship

snusnu committed:
Introduce initial Relationship::Builder code

snusnu committed:
Rename Relationship#builder_class to #mapper_builder_class

snusnu committed:
Make it Relationship#mapper_builder instead

snusnu committed:
Don't mark passing specs as pending

snusnu committed:
Also perform option validation in Builder::Has

snusnu committed:
Remove unused methods from Mapper::RelationshipSet

solnic commented:
@snusnu can we extract Inflector to a separate project?

solnic committed:
Minor simplification in Utils.extract_type

solnic committed:
Remove redundant comma from m:m spec

solnic committed:
Minor formatting fixes

solnic commented:
@snusnu I thought we've decided to use 'parent' and 'child' instead of 'source' and 'target'? Change of plans? :) /cc @dkubb

solnic committed:
Add Relation#inspect

solnic committed:
Add Attribute::Primitive#inspect and Attribute::Mapper#inspect

solnic committed:
Add pending example with nested query hash

solnic commented:
what do you guys think about renaming `Attribute::Mapper` => `Attribute::MapperAttribute`, `Attribute::Collection` => `Attribute::CollectionAttribute` and `Attribute::Primitive` => `Attribute::PrimitiveAttribute`?

I usually don't like adding suffixes to classes that live inside a proper namespace but having `DataMapper::Mapper::Attribute::Mapper` looks awkward.

solnic commented:
@snusnu Heads up - I made some super minor changes. See my commits from today.

snusnu committed:
Demonstrate passing a block to generated mappers

While yielding correct results, I think this is
not the desired API. I'm thinking that the block
that gets passed in, should be evaluated in the
context of the generated mapper.

snusnu commented:
@solnic What about renaming `Mapper::Attribute::Mapper` to `Mapper::Attribute::Model`?

solnic commented:
Hmm yeah that's even better. Or maybe `Attribute::EmbeddedValue` and `Attribute::EmbeddedCollection` kinda like we have in Virtus.

On 22 wrz 2012, at 15:18, Martin Gamsjaeger notifications@github.com wrote:

> @solnic What about renaming Mapper::Attribute::Mapper to Mapper::Attribute::Model?
> 
> —
> Reply to this email directly or view it on GitHub.

snusnu commented:
Hmm, imho the word "Embedded" makes too many assumptions

snusnu commented:
Although looking at it from the object's POV only, it's probably fair to say that both kinds are embedded into it.

mbj commented:
I use EmbeddedDocument and EmbeddedCollection also in my mapper.

On Sat, Sep 22, 2012 at 07:32:42AM -0700, Piotr Solnica wrote:

> Hmm yeah that's even better. Or maybe `Attribute::EmbeddedValue` and `Attribute::EmbeddedCollection` kinda like we have in Virtus.
> 
> On 22 wrz 2012, at 15:18, Martin Gamsjaeger notifications@github.com wrote:
> 
> > @solnic What about renaming Mapper::Attribute::Mapper to Mapper::Attribute::Model?
> > 
> > —
> > Reply to this email directly or view it on GitHub.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/solnic/dm-mapper/pull/6#issuecomment-8789315

## 

Markus Schirp

Phone:   +49 201 / 360 379 14 
Fax:     +49 201 / 360 379 16
Web:     www.seonic.net
Email:   mbj@seonic.net
Twitter: twitter.com/_m_b_j_
OS-Code: github.com/mbj

Seonic IT-Systems GbR
Anton Shatalov & Markus Schirp
Altendorferstrasse 44
D-45127 Essen

snusnu commented:
I think I like `EmbeddedValue` and `EmbeddedCollection` now. `Model` seems to be wrong anyway, as it actually fits better to think of any mapper that users define as a model mapper.

snusnu committed:
Rename EV and EC attribute mappers

Attribute::Mapper     => Attribute::EmbeddedValue
Attribute::Collection => Attribute::EmbeddedCollection

solnic commented:
w00t :)
